# working with MySQL / MariaDB from python
* lesson 13 used `sqlite3` - a lightweight, embedded database that lives in one file (or purely in RAM)
* lesson 42 (running in parallel to this one) covers **PostgreSQL** - a real database *server* process, connected to over the network with `psycopg2` and SQLAlchemy
* this lesson covers the OTHER hugely popular open-source database server: **MySQL** - and specifically **MariaDB**, which is what we are actually connecting to here
* quick bit of history, because it explains why we keep saying "MySQL/MariaDB" together: MySQL was originally developed by a company that was eventually bought by Oracle - a lot of the open-source community got nervous about one single company (especially one with Oracle's reputation around licensing) owning such a widely used piece of infrastructure, so the original MySQL creators forked the codebase into **MariaDB**, a fully open-source project with no single corporate owner
* MariaDB is, for essentially everything we do in this lesson, a **drop-in replacement** for MySQL - same wire protocol, same SQL dialect, same `pymysql`/`mysqlclient` python drivers work against either one unchanged. so "MySQL-compatible" and "speaks the MySQL protocol" is the accurate way to think about MariaDB, and we will just say "MySQL" for most of this lesson to mean "the MySQL family, including MariaDB"
* why does any of this matter historically? MySQL was the "M" in the classic **LAMP stack** (Linux, Apache, MySQL, PHP) - the stack that powered a *huge* chunk of the early web. countless PHP-based websites, forums and content management systems were built on top of it
* it is still everywhere today: **WordPress** (which alone powers something like 40% of all websites on the internet) still defaults to MySQL/MariaDB as its database, and so do many other PHP-based CMSs like Drupal and Joomla
* the good news for us: everything we learned in lesson 13 (tables, rows, primary keys, `cursor.execute()`, parameterized queries, `JOIN`) and lesson 40 (SQLAlchemy's `Engine`/Core/ORM) transfers almost 1:1 here too - what's genuinely new in this lesson is: a different low-level driver (`pymysql`), a handful of real MySQL-specific quirks and features, and (again) how little of our actual python code needs to change to target yet another database engine

## connecting for real: server details
* unlike sqlite3 (which just opens a file), MySQL is a real database **server** process - something has to actually be running and listening on a network port before we can connect at all
* in this sandbox, a MariaDB server is already running on `127.0.0.1` (localhost), port `3306` (MySQL's traditional default port, same idea as postgres's `5432` from lesson 42)
* we already have a database called `pytraining` and a user `pyuser`/`pypass` with full privileges on it - set up for us ahead of time, exactly like lesson 42's postgres setup
* `pymysql` is a pure-python MySQL driver (`pip install pymysql`, already installed here) - there's also `mysqlclient` (a C-extension wrapper around the official MySQL C library, faster but needs compiling) and `mysql-connector-python` (Oracle's own official driver) - all three implement the same DB-API 2.0 shape, so switching between them barely changes your code at all

In [1]:
import pymysql  # the pure-python MySQL/MariaDB driver we installed for this lesson

# connection details for the MariaDB server already running in this sandbox
DB_HOST = "127.0.0.1"      # localhost - the server is running right here
DB_PORT = 3306              # MySQL's traditional default port (postgres uses 5432, recap lesson 42)
DB_USER = "pyuser"          # a user account that was already created for us with full rights on the "pytraining" db
DB_PASSWORD = "pypass"      # NEVER hardcode real passwords like this in real code - use environment variables/secrets managers instead
DB_NAME = "pytraining"      # the database (schema) we are working in

# pymysql.connect() is our equivalent of sqlite3.connect() from lesson 13 - it opens a real network connection this time
connection = pymysql.connect(
    host=DB_HOST,
    port=DB_PORT,
    user=DB_USER,
    password=DB_PASSWORD,
    database=DB_NAME,
)

# just like sqlite3, we get a cursor to actually run SQL and fetch results
cursor = connection.cursor()

# a quick sanity check query, so we know we are really talking to a live MariaDB server and not imagining it
cursor.execute("SELECT VERSION()")
print("connected! server reports version:", cursor.fetchone()[0])

# pymysql also does NOT autocommit by default (same as sqlite3, same as psycopg2 in lesson 42) -
# changes only become permanent once we call connection.commit(), we will see this in action below
print("autocommit is currently:", connection.get_autocommit())

connected! server reports version: 10.11.14-MariaDB-0ubuntu0.24.04.1
autocommit is currently: False


## creating a table and inserting rows - and a placeholder-style gotcha
* the pattern is IDENTICAL to lesson 13's sqlite3: `cursor.execute(sql, params)`, `cursor.executemany(sql, list_of_params)`, `connection.commit()`
* here is a genuinely annoying inconsistency worth knowing about: sqlite3 uses `?` as its placeholder character, but `pymysql` (and psycopg2 in lesson 42!) use `%s` instead - even though we are NOT doing old-style python `%`-formatting, `%s` here is just pymysql's chosen placeholder syntax, values are still passed separately and still safely escaped
* this is exactly the kind of small-but-real friction that comes from every python database driver inventing its own placeholder style - and it's a big part of why SQLAlchemy (which we get to right after this) is so popular: it hides this exact detail behind one consistent API
* we prefix every table we create in this lesson with `l43_` - just a good habit to avoid clashing with anything else that might be using this shared database, and we `DROP` them all again at the very end

In [2]:
# creating our first table - note the l43_ prefix, and that this is completely normal SQL, no python-object magic yet
cursor.execute('''
    CREATE TABLE IF NOT EXISTS l43_authors (
        id INT AUTO_INCREMENT PRIMARY KEY,   -- MySQL's way of saying "auto-increment this for me", more on this below
        name VARCHAR(100) NOT NULL           -- MySQL wants a length for VARCHAR, unlike sqlite3's untyped TEXT
    )
''')
connection.commit()  # DDL (CREATE TABLE) doesnt strictly need a commit in MySQL (it auto-commits), but it doesnt hurt to be explicit
print("l43_authors table created")

# inserting with %s placeholders - THIS is the pymysql/psycopg2 style, different from sqlite3's "?" (recap lesson 13)
cursor.execute(
    "INSERT INTO l43_authors (name) VALUES (%s)",  # %s is the placeholder here, NOT a python %-format string
    ("Ursula K. Le Guin",),                          # the actual value, passed separately - still 100% safe from SQL injection
)
connection.commit()
print("inserted one author, new id was:", cursor.lastrowid)  # more on lastrowid in the very next section

# executemany works exactly like sqlite3's version - one call, many rows
more_authors = [
    ("Isaac Asimov",),
    ("Terry Pratchett",),
]
cursor.executemany("INSERT INTO l43_authors (name) VALUES (%s)", more_authors)
connection.commit()
print("authors table now has", end=" ")

cursor.execute("SELECT COUNT(*) FROM l43_authors")
print(cursor.fetchone()[0], "rows")

# querying back - .fetchall() works exactly like lesson 13's sqlite3 version
cursor.execute("SELECT id, name FROM l43_authors ORDER BY id")
for row in cursor.fetchall():
    print(" -", row)

l43_authors table created
inserted one author, new id was: 1
authors table now has 3 rows
 - (1, 'Ursula K. Le Guin')
 - (2, 'Isaac Asimov')
 - (3, 'Terry Pratchett')


## the same thing again, but with SQLAlchemy
* recap lesson 40: SQLAlchemy's whole pitch is that Core/ORM code barely changes when you swap database engines - the connection URL is basically the only thing that's different
* the URL shape is `dialect+driver://user:password@host:port/database` - for us that's `mysql+pymysql://...` (mysql = the dialect/SQL flavor, pymysql = which actual driver package to use underneath)
* everything else - `Table`/`Column`/`MetaData`, `insert()`/`select()`, `Session`, `Mapped`/`mapped_column` - is *exactly* the same code we would write against sqlite (lesson 40) or postgres (lesson 42)
* this drives home the earlier point about `%s` vs `?`: SQLAlchemy generates the right placeholder style for whichever driver is underneath it, so we never even have to think about it once we're working through Core/the ORM

In [3]:
from sqlalchemy import create_engine, MetaData, Table, Column, Integer, String, insert, select

# mysql+pymysql:// - "mysql" is the dialect (which SQL flavor to speak), "pymysql" is which driver package to actually use
engine = create_engine("mysql+pymysql://pyuser:pypass@127.0.0.1:3306/pytraining")
print("engine dialect:", engine.dialect.name, "| driver:", engine.dialect.driver)

metadata = MetaData()

# describing a table as python objects, exactly like lesson 40 - note we again use the l43_ prefix
l43_products_core = Table(
    "l43_products_core",
    metadata,
    Column("id", Integer, primary_key=True),          # SQLAlchemy translates this into an AUTO_INCREMENT column for MySQL specifically
    Column("name", String(100), nullable=False),
    Column("price_cents", Integer, nullable=False),     # storing money as integer cents avoids floating point rounding issues
)

metadata.create_all(engine)  # this actually issues a real CREATE TABLE against our MariaDB server
print("table created via Core:", list(metadata.tables.keys()))

with engine.connect() as connection_sa:
    # insert() + .values() - completely identical code to lesson 40's sqlite example
    result = connection_sa.execute(
        insert(l43_products_core),
        [
            {"name": "Coffee Mug", "price_cents": 1299},
            {"name": "Notebook", "price_cents": 499},
        ],
    )
    connection_sa.commit()

    # select() + iterate - again, identical to lesson 40, no MySQL-specific python code required at all
    rows = connection_sa.execute(select(l43_products_core).order_by(l43_products_core.c.price_cents)).all()
    print("products, cheapest first:")
    for row in rows:
        print(" -", row.name, f"(${row.price_cents / 100:.2f})")

engine dialect: mysql | driver: pymysql
table created via Core: ['l43_products_core']
products, cheapest first:
 - Notebook ($4.99)
 - Coffee Mug ($12.99)


## MySQL-specific thing #1: AUTO_INCREMENT and no RETURNING clause
* sqlite3 gives us auto-increment "for free" on an `INTEGER PRIMARY KEY` column, postgres calls its version `SERIAL`/`IDENTITY` (lesson 42) - MySQL's name for the same concept is `AUTO_INCREMENT`, which we already used above
* after an insert, we usually want to know the id the database just generated for us - with raw `pymysql` that's `cursor.lastrowid` (or equivalently `connection.insert_id()`), exactly the same *idea* as sqlite3's `cursor.lastrowid` from lesson 13
* here's a real, sometimes-annoying difference from postgres: MySQL has **no `RETURNING` clause**. in postgres (lesson 42) you can do `INSERT ... RETURNING id, some_other_column` and get arbitrary columns back from the very same statement. MySQL simply cannot do that (as of the versions in wide use today) - `lastrowid` only ever gives you the auto-increment id, and if you need any OTHER column's value (say, a column with a default computed by the database), you have to run a separate `SELECT` afterwards
* this is a genuine portability wrinkle: code that leans on postgres's `RETURNING` for anything beyond the id needs a rewrite (an extra round-trip `SELECT`) to run against MySQL

In [4]:
# lastrowid after a plain pymysql insert - this is MySQL's answer to "give me the id you just generated"
cursor.execute("INSERT INTO l43_authors (name) VALUES (%s)", ("Frank Herbert",))
# connection.insert_id() is pymysql's equivalent way to ask the same question - but it has to be read BEFORE
# we commit: issuing COMMIT is itself another round trip to the server, and that resets the tracked value back to 0
same_id_via_connection = connection.insert_id()
connection.commit()
new_id = cursor.lastrowid  # cursor.lastrowid stays correct even after commit - it's cached client-side, more idiomatic DB-API style
print("new author id via cursor.lastrowid:", new_id)
print("same value via connection.insert_id() (read before the commit):", same_id_via_connection)

# what we CANNOT do in MySQL: something like postgres's "INSERT ... RETURNING name" to get another column back in one round trip
# instead, if we need more than just the id, we have to go fetch it ourselves afterwards:
cursor.execute("SELECT name FROM l43_authors WHERE id = %s", (new_id,))
print("had to run a SEPARATE select to confirm the name:", cursor.fetchone()[0])

# SQLAlchemy smooths this over a little bit for the common "just give me the new id" case -
# result.inserted_primary_key works the same way across sqlite/postgres/mysql even though MySQL has no RETURNING under the hood
with engine.connect() as connection_sa:
    result = connection_sa.execute(insert(l43_products_core).values(name="Desk Lamp", price_cents=2499))
    connection_sa.commit()
    print("SQLAlchemy still hands us the new id portably:", result.inserted_primary_key[0])

new author id via cursor.lastrowid: 4
same value via connection.insert_id() (read before the commit): 4
had to run a SEPARATE select to confirm the name: Frank Herbert
SQLAlchemy still hands us the new id portably: 3


## MySQL-specific thing #2: storage engines - InnoDB vs MyISAM
* MySQL is unusual among the big relational databases in that it has **pluggable storage engines** - the part that actually decides how table data is physically stored and how transactions/locking work under the hood
* **InnoDB** is the default engine today (has been for many years) - it's fully **transactional** (supports `COMMIT`/`ROLLBACK` properly), supports **foreign keys**, and handles concurrent access safely with row-level locking
* **MyISAM** is the older, historical default engine - it does NOT support real transactions (no rollback!) and does NOT support foreign key constraints at all. it used to be chosen for being slightly faster for read-heavy workloads with no concurrent writes, but that tradeoff is basically never worth it today
* the practical takeaway: **always use InnoDB** unless you have an extremely specific, well-understood reason not to (it's the default when you don't specify one, but it's worth knowing this exists, especially if you ever inherit an old codebase with MyISAM tables lying around)
* we can literally ask the database which engine a table uses, via `information_schema` (a special set of "meta" tables every MySQL server exposes, describing its own tables/columns/etc) or the simpler `SHOW TABLE STATUS`

In [5]:
# lets deliberately create one table on each engine, purely to demonstrate the difference
cursor.execute("CREATE TABLE IF NOT EXISTS l43_innodb_demo (id INT PRIMARY KEY) ENGINE=InnoDB")
cursor.execute("CREATE TABLE IF NOT EXISTS l43_myisam_demo (id INT PRIMARY KEY) ENGINE=MyISAM")
connection.commit()
print("created one InnoDB table and one MyISAM table")

# information_schema.TABLES lets us ask any MySQL server "what engine is this table using?" - very handy for auditing an existing db
cursor.execute('''
    SELECT TABLE_NAME, ENGINE
    FROM information_schema.TABLES
    WHERE TABLE_SCHEMA = %s AND TABLE_NAME IN (%s, %s)
''', ("pytraining", "l43_innodb_demo", "l43_myisam_demo"))
print("engines according to information_schema:")
for table_name, engine_name in cursor.fetchall():
    print(" -", table_name, "->", engine_name)

# SHOW TABLE STATUS is another, more MySQL-idiomatic way to see the same thing (plus row counts, size, etc)
cursor.execute("SHOW TABLE STATUS WHERE Name = 'l43_innodb_demo'")
status_row = cursor.fetchone()
# SHOW TABLE STATUS returns a bunch of columns - Name is index 0, Engine is index 1
print(f"SHOW TABLE STATUS for l43_innodb_demo: Name={status_row[0]!r}, Engine={status_row[1]!r}")

# why InnoDB matters in practice: it actually respects ROLLBACK. lets prove it right here with the MyISAM table as a contrast
connection.begin()  # start an explicit transaction on the InnoDB-backed default table (l43_authors uses InnoDB, MySQL's default)
cursor.execute("INSERT INTO l43_authors (name) VALUES (%s)", ("Will Be Rolled Back",))
connection.rollback()  # throw the whole transaction away
cursor.execute("SELECT COUNT(*) FROM l43_authors WHERE name = %s", ("Will Be Rolled Back",))
print("rows matching after rollback (should be 0, InnoDB honored it):", cursor.fetchone()[0])

# cleaning up our two little demo tables straight away, we only needed them to show the engine difference
cursor.execute("DROP TABLE l43_innodb_demo")
cursor.execute("DROP TABLE l43_myisam_demo")
connection.commit()

created one InnoDB table and one MyISAM table
engines according to information_schema:
 - l43_innodb_demo -> InnoDB
 - l43_myisam_demo -> MyISAM
SHOW TABLE STATUS for l43_innodb_demo: Name='l43_innodb_demo', Engine='InnoDB'
rows matching after rollback (should be 0, InnoDB honored it): 0


## MySQL-specific thing #3: case-insensitive string comparisons (a classic gotcha!)
* here's a genuinely surprising difference if you're coming from sqlite3 or postgres: by default, MySQL's most common text column type (`VARCHAR`) uses a **case-insensitive** collation (`utf8mb4_general_ci`/`utf8mb4_0900_ai_ci` depending on version - the `_ci` literally stands for "case insensitive")
* that means `WHERE name = 'ursula'` can match a row where the actual stored value is `'URSULA'` or `'Ursula'` - something that would NOT happen in sqlite3 or postgres, where text equality is case-SENSITIVE by default
* this is a classic real-world "gotcha" when porting an application between database engines: code and tests that quietly relied on MySQL's case-insensitive matching can start failing in confusing ways once ported to postgres (or vice versa: code relying on case-sensitivity can start returning EXTRA, unwanted rows once ported to MySQL)
* you CAN control this explicitly with `COLLATE` (e.g. `utf8mb4_bin` for byte-exact, case-sensitive comparisons) - but the crucial thing to know is what your database's *default* actually does, since most day to day queries dont specify a collation explicitly at all

In [6]:
# lets prove this concretely - we already have "Ursula K. Le Guin" stored with mixed case
cursor.execute("SELECT name FROM l43_authors WHERE name = %s", ("URSULA K. LE GUIN",))  # note: ALL CAPS in our query
matched = cursor.fetchall()
print("querying in ALL CAPS still matched:", matched)  # this WOULD return zero rows in sqlite3 or postgres by default!

# same thing again but fully lowercase this time, just to hammer the point home
cursor.execute("SELECT name FROM l43_authors WHERE name = %s", ("terry pratchett",))
print("querying in all-lowercase still matched:", cursor.fetchall())

# we can check the actual collation MySQL is using for this column via information_schema
cursor.execute('''
    SELECT COLLATION_NAME
    FROM information_schema.COLUMNS
    WHERE TABLE_SCHEMA = %s AND TABLE_NAME = %s AND COLUMN_NAME = %s
''', ("pytraining", "l43_authors", "name"))
print("actual collation of l43_authors.name:", cursor.fetchone()[0])

# and we CAN force case-sensitive comparison with an explicit COLLATE if we ever need it
cursor.execute("SELECT name FROM l43_authors WHERE name COLLATE utf8mb4_bin = %s", ("URSULA K. LE GUIN",))
print("with COLLATE utf8mb4_bin (case-sensitive) forced, ALL CAPS matches:", cursor.fetchall())  # now correctly finds nothing

querying in ALL CAPS still matched: (('Ursula K. Le Guin',),)
querying in all-lowercase still matched: (('Terry Pratchett',),)
actual collation of l43_authors.name: utf8mb4_general_ci
with COLLATE utf8mb4_bin (case-sensitive) forced, ALL CAPS matches: ()


## MySQL-specific thing #4: the JSON column type
* just like postgres's `JSONB` (lesson 42), MySQL (5.7+) and MariaDB (10.2+) have a native `JSON` column type - it validates that whatever you insert is actually valid JSON
* querying INTO the json is where the syntax differs from postgres: the portable, works-everywhere-in-the-MySQL-family function is `JSON_EXTRACT(column, path)`, and `JSON_UNQUOTE(JSON_EXTRACT(column, path))` if we want the value back as plain text instead of a quoted json string - the path syntax `'$.field'` always starts with `$` (meaning "the document root")
* real MySQL also offers a shorthand: the `->` operator (same as `JSON_EXTRACT`) and `->>` operator (same as `JSON_UNQUOTE(JSON_EXTRACT(...))`) - **but here's a genuinely interesting, real MariaDB-vs-MySQL divergence we bumped into while writing this very lesson**: the MariaDB server we're actually connected to (10.11) does NOT understand `->`/`->>` at all - it errors out with a plain syntax error! MariaDB implements JSON as a `LONGTEXT` column with a validity `CHECK` constraint under the hood (not MySQL's dedicated binary JSON storage format), and it never picked up MySQL's `->`/`->>` shorthand syntax on top of it
* so unlike almost everything else in this lesson, this is one spot where "MySQL and MariaDB are drop-in compatible" genuinely breaks down a little - `JSON_EXTRACT()`/`JSON_UNQUOTE()` work identically on both, but code using the `->>` shorthand (very common in MySQL tutorials/blog posts!) will simply fail here. the safe, portable choice across the whole MySQL/MariaDB family is to stick to the explicit functions, which is exactly what we do below
* postgres's JSONB (lesson 42) has its own `->`/`->>` operators plus a rich set of containment/indexing operators - the CONCEPT (store semi-structured data in a normally-relational column, query into specific fields) is shared everywhere, the exact syntax is engine (and, as we just saw, even engine-*variant*) specific

In [7]:
# a table with a native JSON column - useful for e.g. "extra, variable-shape metadata" that doesnt deserve its own columns
cursor.execute('''
    CREATE TABLE IF NOT EXISTS l43_events (
        id INT AUTO_INCREMENT PRIMARY KEY,
        event_type VARCHAR(50) NOT NULL,
        payload JSON NOT NULL          -- MySQL validates this is actually well-formed JSON at insert time
    )
''')
connection.commit()

# two rows with genuinely DIFFERENT json "shapes" - this is exactly the kind of flexibility a JSON column is good for
cursor.execute(
    "INSERT INTO l43_events (event_type, payload) VALUES (%s, %s)",
    ("login", '{"user_id": 42, "ip": "10.0.0.5"}'),
)
cursor.execute(
    "INSERT INTO l43_events (event_type, payload) VALUES (%s, %s)",
    ("purchase", '{"user_id": 42, "items": ["mug", "notebook"], "total_cents": 1798}'),
)
connection.commit()
print("inserted 2 events with different json shapes")

# JSON_EXTRACT() pulls a specific field out - the path always starts with $ (the document root)
cursor.execute("SELECT event_type, JSON_EXTRACT(payload, '$.user_id') FROM l43_events")
print("using JSON_EXTRACT:")
for event_type, user_id in cursor.fetchall():
    print(" -", event_type, "user_id =", user_id)

# JSON_UNQUOTE(JSON_EXTRACT(...)) does the same extraction, but unwraps it as plain text (no surrounding quotes) -
# this is the portable way to write what MySQL's ->> shorthand does, and it also works fine on MariaDB (recap the markdown above)
cursor.execute(
    "SELECT event_type, JSON_UNQUOTE(JSON_EXTRACT(payload, '$.user_id')) AS user_id FROM l43_events WHERE event_type = %s",
    ("purchase",),
)
print("using JSON_UNQUOTE(JSON_EXTRACT(...)):", cursor.fetchone())

# we can even query INTO an array thats nested inside the json, using an index in the path
cursor.execute(
    "SELECT JSON_UNQUOTE(JSON_EXTRACT(payload, '$.items[0]')) AS first_item FROM l43_events WHERE event_type = %s",
    ("purchase",),
)
print("first item of the purchase's items array:", cursor.fetchone()[0])

inserted 2 events with different json shapes
using JSON_EXTRACT:
 - login user_id = 42
 - purchase user_id = 42
using JSON_UNQUOTE(JSON_EXTRACT(...)): ('purchase', '42')
first item of the purchase's items array: mug


## transactions with pymysql: explicit START TRANSACTION / COMMIT / ROLLBACK
* recap lesson 13/42: a transaction groups multiple statements so they either ALL succeed together, or none of them do
* pymysql, like psycopg2 (lesson 42), does NOT autocommit by default - every statement is already implicitly inside an open transaction the moment we start executing, we just need to actually call `.commit()` (or `.rollback()`) to end it
* `connection.begin()` (what we used above) is pymysql's python-level way to explicitly start one; we can also just execute the raw SQL `START TRANSACTION` if we want to be fully explicit at the SQL level, which is what we'll do here
* the classic real-world use case: a multi-step operation (e.g. "move money from account A to account B") where we absolutely do NOT want to leave things half-done if step 2 fails after step 1 already succeeded - `ROLLBACK` undoes EVERYTHING since the transaction started

In [8]:
# a tiny "bank accounts" table to demonstrate a realistic rollback scenario
cursor.execute('''
    CREATE TABLE IF NOT EXISTS l43_accounts (
        id INT AUTO_INCREMENT PRIMARY KEY,
        owner VARCHAR(100) NOT NULL,
        balance_cents INT NOT NULL
    )
''')
connection.commit()

cursor.execute(
    "INSERT INTO l43_accounts (owner, balance_cents) VALUES (%s, %s), (%s, %s)",
    ("Alice", 10000, "Bob", 5000),  # Alice starts with $100.00, Bob with $50.00
)
connection.commit()
print("starting balances:")
cursor.execute("SELECT owner, balance_cents FROM l43_accounts ORDER BY owner")
print(cursor.fetchall())

# scenario: we try to transfer $30.00 from Alice to Bob, but something goes wrong partway through
cursor.execute("START TRANSACTION")  # explicit SQL-level transaction start
try:
    cursor.execute("UPDATE l43_accounts SET balance_cents = balance_cents - %s WHERE owner = %s", (3000, "Alice"))
    print("step 1 done: subtracted $30.00 from Alice (not committed yet!)")

    # simulate something going wrong before we get to credit Bob - e.g. a bug, a network blip, a failed validation
    raise RuntimeError("simulated failure before crediting Bob!")

    # this line would have been step 2, but we never get here because of the exception above:
    # cursor.execute("UPDATE l43_accounts SET balance_cents = balance_cents + %s WHERE owner = %s", (3000, "Bob"))
except RuntimeError as error:
    print("caught an error mid-transaction:", error)
    connection.rollback()  # THIS is the important line - undoes the (uncommitted) subtraction from Alice too
    print("rolled back the whole transaction")

# verify: Alice's balance should be completely UNCHANGED, the partial update never became permanent
cursor.execute("SELECT owner, balance_cents FROM l43_accounts ORDER BY owner")
print("balances after rollback (Alice should still have 10000):", cursor.fetchall())

starting balances:
(('Alice', 10000), ('Bob', 5000))
step 1 done: subtracted $30.00 from Alice (not committed yet!)
caught an error mid-transaction: simulated failure before crediting Bob!
rolled back the whole transaction
balances after rollback (Alice should still have 10000): (('Alice', 10000), ('Bob', 5000))


## cleanup: dropping every l43_ table we created
* we prefixed everything with `l43_` specifically so we can find and drop it all again cleanly at the end, leaving the shared `pytraining` database exactly as we found it
* `information_schema.TABLES` (which we already used above to inspect engines) is a handy way to programmatically find every table matching our prefix, rather than hardcoding a list and risking forgetting one

In [9]:
# find every table we created in this lesson, by our l43_ prefix - programmatically, so we cant forget one
cursor.execute('''
    SELECT TABLE_NAME FROM information_schema.TABLES
    WHERE TABLE_SCHEMA = %s AND TABLE_NAME LIKE %s
''', ("pytraining", "l43_%"))
our_tables = [row[0] for row in cursor.fetchall()]
print("l43_ tables found before cleanup:", our_tables)

# drop every single one of them
for table_name in our_tables:
    cursor.execute(f"DROP TABLE {table_name}")  # safe here: table_name came from information_schema, not from any user input
connection.commit()
print("dropped", len(our_tables), "tables")

# double check with the exact same query - this should now come back empty
cursor.execute('''
    SELECT TABLE_NAME FROM information_schema.TABLES
    WHERE TABLE_SCHEMA = %s AND TABLE_NAME LIKE %s
''', ("pytraining", "l43_%"))
remaining = cursor.fetchall()
print("l43_ tables remaining after cleanup:", remaining)  # should be an empty tuple/list - nothing left behind

# the raw pymysql loop above already dropped l43_products_core too (it matched the l43_ prefix via information_schema)
# metadata.drop_all() checks what actually still exists before issuing DROP TABLE, so calling it here is a safe no-op -
# still good practice to call it, in case any Core-managed table had NOT already been caught by the loop above
metadata.drop_all(engine)
print("called metadata.drop_all() too, just in case (no-op here, the loop above already got everything)")

# closing both connections properly now that we are completely done
cursor.close()
connection.close()
engine.dispose()  # tells SQLAlchemy to close every pooled connection it was holding onto
print("connections closed")

l43_ tables found before cleanup: ['l43_events', 'l43_accounts', 'l43_products_core', 'l43_authors']
dropped 4 tables
l43_ tables remaining after cleanup: ()
called metadata.drop_all() too, just in case (no-op here, the loop above already got everything)
connections closed


## quick comparison to other languages
* just like postgres (lesson 42), MySQL/MariaDB has a driver for basically every language: Java's **MySQL Connector/J** (JDBC-based), Node's **`mysql2`** package, C#'s **`MySqlConnector`**, Go's **`go-sql-driver/mysql`**, Rust's **`sqlx`** or **`mysql`** crates - all of them shaped around their own language's usual database idiom (JDBC's `Connection`/`Statement`/`ResultSet` for Java, DB-API-style for python, etc)
* the practical postgres-vs-MySQL choice comes up constantly on real teams, and both are excellent, free, mature options today - but they do have different reputations that often (still) hold true in practice:
  * MySQL/MariaDB has historically had a reputation for being slightly simpler to operate/tune for straightforward workloads, and it is EXTREMELY widely deployed - virtually every shared web host and one-click-installer supports it, and it's the default choice for WordPress, Drupal and most PHP CMSs
  * PostgreSQL has a reputation for being more feature-rich and standards-compliant for advanced use cases - lesson 42 showed off `JSONB`, arrays, window functions and `RETURNING`, several of which either don't exist in MySQL at all or work quite differently (as we saw with `RETURNING` in this very lesson)
* in practice today, the "right" choice for a new project often comes down to team familiarity, what your hosting provider supports well, or one specific feature tipping the scale (postgres's `JSONB`/arrays/PostGIS geospatial extension being common tie-breakers toward postgres) - both are genuinely solid, boring-in-the-good-way choices, and knowing both (as we now do, across lessons 13/40/42/43) means you're never stuck if a job/project has already picked one for you

## Exercises
1. create an `l43_products` table (id, name, price_cents) and insert one row with plain `pymysql` - use `cursor.lastrowid` right after the insert to get the new row's id WITHOUT running a separate `SELECT`
2. insert an author name in one particular case (e.g. `"Grace Hopper"`) into a fresh `l43_people` table, then write a query of your own design that matches it using a DIFFERENT case (e.g. searching for `"GRACE HOPPER"`) to demonstrate MySQL's default case-insensitive string comparison
3. create an `l43_products_meta` table with an `id` and a `JSON` column called `attributes`, insert 2 rows with genuinely different JSON shapes (e.g. one product with a `"color"` field, another with a `"weight_grams"` field instead), then query one specific field out of each using `JSON_EXTRACT()`/`JSON_UNQUOTE()`
4. write your own scripted rollback scenario (similar in spirit to the bank account transfer above, but come up with your own story - e.g. two `UPDATE`s to an inventory table where the second one should never happen): start a transaction, make one change, deliberately trigger a `ROLLBACK` instead of committing, then verify with a `SELECT` that nothing was actually saved

In [10]:
# TODO: solve exercise 1 here (l43_products table, insert one row, get id via cursor.lastrowid)


# TODO: solve exercise 2 here (l43_people table, insert one case, query with a different case)


# TODO: solve exercise 3 here (l43_products_meta table with JSON column, 2 differently-shaped rows, query with JSON_EXTRACT/->>)


# TODO: solve exercise 4 here (your own transaction + rollback scenario, verify nothing was saved)

### sample solutions
* try the exercises above yourself first, only look here if you are stuck or want to compare your approach
* we open one fresh connection here for the solutions, and clean up (drop tables, close connection) again right at the end

In [11]:
import pymysql

# a fresh connection for the exercise solutions, same details as at the top of the lesson
solutions_connection = pymysql.connect(host="127.0.0.1", port=3306, user="pyuser", password="pypass", database="pytraining")
solutions_cursor = solutions_connection.cursor()

# --- sample solution 1: l43_products + cursor.lastrowid, no separate SELECT needed for the id ---
solutions_cursor.execute('''
    CREATE TABLE IF NOT EXISTS l43_products (
        id INT AUTO_INCREMENT PRIMARY KEY,
        name VARCHAR(100) NOT NULL,
        price_cents INT NOT NULL
    )
''')
solutions_cursor.execute("INSERT INTO l43_products (name, price_cents) VALUES (%s, %s)", ("Keyboard", 7999))
solutions_connection.commit()
new_product_id = solutions_cursor.lastrowid  # this IS the point of the exercise - no extra SELECT required
print("exercise 1 - new product id straight from lastrowid:", new_product_id)


# --- sample solution 2: case-insensitive matching demo, own table/name ---
solutions_cursor.execute('''
    CREATE TABLE IF NOT EXISTS l43_people (
        id INT AUTO_INCREMENT PRIMARY KEY,
        full_name VARCHAR(100) NOT NULL
    )
''')
solutions_cursor.execute("INSERT INTO l43_people (full_name) VALUES (%s)", ("Grace Hopper",))  # stored in mixed case
solutions_connection.commit()

# now query for the SAME person, but typed in a totally different case - MySQL's default collation matches it anyway
solutions_cursor.execute("SELECT full_name FROM l43_people WHERE full_name = %s", ("GRACE HOPPER",))
print("exercise 2 - matched despite different case:", solutions_cursor.fetchall())


# --- sample solution 3: JSON column with two different shapes, queried with JSON_EXTRACT / ->> ---
solutions_cursor.execute('''
    CREATE TABLE IF NOT EXISTS l43_products_meta (
        id INT AUTO_INCREMENT PRIMARY KEY,
        attributes JSON NOT NULL
    )
''')
solutions_cursor.execute("INSERT INTO l43_products_meta (attributes) VALUES (%s)", ('{"color": "red"}',))
solutions_cursor.execute("INSERT INTO l43_products_meta (attributes) VALUES (%s)", ('{"weight_grams": 350}',))
solutions_connection.commit()

solutions_cursor.execute(
    "SELECT id, JSON_EXTRACT(attributes, '$.color') FROM l43_products_meta "
    "WHERE JSON_UNQUOTE(JSON_EXTRACT(attributes, '$.color')) IS NOT NULL"
)
print("exercise 3 - row with a color field:", solutions_cursor.fetchone())
solutions_cursor.execute(
    "SELECT id, JSON_UNQUOTE(JSON_EXTRACT(attributes, '$.weight_grams')) FROM l43_products_meta "
    "WHERE JSON_UNQUOTE(JSON_EXTRACT(attributes, '$.weight_grams')) IS NOT NULL"
)
print("exercise 3 - row with a weight_grams field:", solutions_cursor.fetchone())


# --- sample solution 4: our own rollback scenario - an inventory "reserve stock" that should NOT go through ---
solutions_cursor.execute('''
    CREATE TABLE IF NOT EXISTS l43_inventory (
        id INT AUTO_INCREMENT PRIMARY KEY,
        item VARCHAR(100) NOT NULL,
        quantity INT NOT NULL
    )
''')
solutions_cursor.execute("INSERT INTO l43_inventory (item, quantity) VALUES (%s, %s)", ("Widgets", 100))
solutions_connection.commit()

solutions_cursor.execute("START TRANSACTION")
solutions_cursor.execute("UPDATE l43_inventory SET quantity = quantity - %s WHERE item = %s", (20, "Widgets"))
print("exercise 4 - reserved 20 widgets (uncommitted)")
# imagine here we discover the order actually got cancelled before we could commit - so we bail out instead
solutions_connection.rollback()
solutions_cursor.execute("SELECT quantity FROM l43_inventory WHERE item = %s", ("Widgets",))
print("exercise 4 - quantity after rollback (should still be 100):", solutions_cursor.fetchone()[0])


# cleanup for the solutions - drop everything we made here too, and close this connection
for table_name in ("l43_products", "l43_people", "l43_products_meta", "l43_inventory"):
    solutions_cursor.execute(f"DROP TABLE IF EXISTS {table_name}")
solutions_connection.commit()

# final check - make absolutely sure NO l43_ table is left anywhere in the database, from either part of this lesson
solutions_cursor.execute('''
    SELECT TABLE_NAME FROM information_schema.TABLES
    WHERE TABLE_SCHEMA = %s AND TABLE_NAME LIKE %s
''', ("pytraining", "l43_%"))
print("final check - any l43_ tables remaining anywhere:", solutions_cursor.fetchall())

solutions_cursor.close()
solutions_connection.close()
print("solutions connection closed")

exercise 1 - new product id straight from lastrowid: 1
exercise 2 - matched despite different case: (('Grace Hopper',),)
exercise 3 - row with a color field: (1, '"red"')
exercise 3 - row with a weight_grams field: (2, '350')
exercise 4 - reserved 20 widgets (uncommitted)
exercise 4 - quantity after rollback (should still be 100): 100
final check - any l43_ tables remaining anywhere: ()
solutions connection closed


* congratulation you finished this lesson

## what we learned
* MySQL/MariaDB is the other major open-source relational database server, historically the "M" in the LAMP stack, still powering huge swaths of the web (WordPress and friends) - MariaDB is a fully open-source, wire-compatible fork/drop-in-replacement born out of concerns about Oracle's ownership of MySQL
* connecting with `pymysql` (DB-API 2.0, exactly like sqlite3 from lesson 13), and that its placeholder style is `%s`, NOT sqlite3's `?` - a small but real inconsistency across python database drivers
* the exact same SQLAlchemy Core/ORM code from lesson 40 works completely unchanged here - only the connection URL (`mysql+pymysql://...`) differs, which is the whole point of using SQLAlchemy in the first place
* `AUTO_INCREMENT` and `cursor.lastrowid`/`connection.insert_id()`, and that MySQL has NO `RETURNING` clause the way postgres does (lesson 42) - a genuine portability wrinkle to watch for
* storage engines: `InnoDB` (transactional, supports foreign keys, the sane modern default) vs the historical `MyISAM` (no transactions, no foreign keys) - and how to check which engine a table uses via `information_schema`/`SHOW TABLE STATUS`
* MySQL's default CASE-INSENSITIVE string comparisons (a classic gotcha when porting between database engines), and how to force case-sensitivity with `COLLATE`
* MySQL's native `JSON` column type, and querying into it with `JSON_EXTRACT()`/`JSON_UNQUOTE()` (conceptually similar to, but syntactically different from, postgres's JSONB in lesson 42) - and the real, hands-on-discovered divergence that MySQL's `->`/`->>` shorthand operators don't even parse on MariaDB, unlike almost everything else in this lesson
* explicit transactions with `START TRANSACTION`/`.commit()`/`.rollback()`, and a concrete rollback scenario undoing a partial multi-statement change
* how MySQL/MariaDB compares across languages (JDBC MySQL Connector, `mysql2`, `MySqlConnector`, `go-sql-driver/mysql`, `sqlx`) and the practical, real-world postgres-vs-MySQL tradeoffs teams weigh today